Architecture diagram and description of the project's folders (#3053)
MirServer/mir: doc/sphinx/architecture.md

# Overview
This document describes the high-level architecture of *Mir*, including a diagram and a brief introduction of the directory structure.

# Concepts
## MirAL

> - The "Mir Abstraction Layer" (aka `mirAL` or `miral`) is a layer that makes the core Mir implementation easier for compositor-implementers to work with.

Reviewer: I don't think this clarifies anything
```suggestion
- The "Mir Abstraction Layer" (aka `MirAL` or `miral`) is an API that makes Mir easy for compositor-implementers to work with. It provides the core window management concepts and an interface that is more ABI stable than Mir's internal mirserver API.
```

> - If one wishes to implement a compositor themselves, they will spend most of their time interacting with the `mirAL` API and only really touch the `mir` API when they require core concepts like a `mir::geometry::Rectangle`.

Reviewer: `Rectangle` is from mircore, and that's very different from the mirserver APIs (that compositor writers should never touch.

- Unlike the `Shell`, the `Scene` knows a lot about the Z-order of the surfaces. For this reason, it is also responsible for things like locking the display or sending surfaces to the back of the Z-order.

## `/src/server/graphics`
- A graphics abstraction layer sitting between the compositor and the specific graphics platform that it is using to render.

## `/src/server/console`
- Handles `logind` and virtual-terminal related tasks for the compositor

## `/src/server/report`
- Logging and other facilities to collect data about the compositor

## `/src/platforms` (with an "s"!)
- Provides both input and rendering platforms that the compositor can use depending on the environment.
- For the rendering platforms, we have:
  - *GBM/KMS*: the platform that you would expect if you were running a *Mir*-based compositor as your daily-driver
  - *X11*: a useful platform for testing locally
  - *Wayland*: another useful platform for testing locally
  - *EGLStreams/KMS*: like the *GBM/KMS* platform, but for Nvidia 
- For the input platforms, we have:
  - evdev
  - X11
  - Wayland

## `/src/platform` (no "s")

> - Graphics and input code that is shared between platforms

Reviewer: Comment: In principle, we aim to make this a public interface for third party "platforms". As Chris has been reworking these APIs, they are currently private

# Overview
This document describes the high-level architecture of *Mir*, including a diagram and a brief introduction of the libraries.

# Audience
This document is intended to provide contributors to *Mir* an overview of *Mir*'s systems. It is *not* intended to guide compositor-implementers, unless they are specifically interested in the internals of *Mir*.

# Diagram
![Architecture Diagram](./architecture.png "Architecture Diagram")

# Top-Level Folders
## `/include`
- The `include` folder at the root of the project provides the public-facing API.
- This directory *strictly* includes the interfaces that are useful to a compositor-implementer.
- If one is only interested in implementing their own *Mir*-based compositor, they need not look any further than this directory.

## `/src`
- The `src` folder at the root of the project comprises the core implementation of *Mir*.

## `/examples`
- A collection of demo *Mir* compositors that are used as examples of how one might build a true compositor. Examples include:
  - `miral-app`

>   - `miral-shel`

Reviewer: ```suggestion
  - `miral-shell`
```

  - `mir_demo_server`

## `/tests`

> - Unit tests for the entire project.

Reviewer: Not just _unit_ tests

# Libraries
The following is the library tree:
```
├── miral
├── miroil
├── mircore
├── mircommon
├── mircookie
└── mirserver
    ├── mircompositor
    ├── mirinput
    ├── mirfrontend
    ├── mirfrontend-wayland
    ├── mirfrontend-xwayland
    ├── mirshell
    ├── mirshelldecoration
    ├── mirscene
    ├── mirgraphics
    ├── mirreport
    ├── mirlttng
    ├── mirlogging
    ├── mirnullreport
    ├── mirconsole
    ├── mirgl

>     └── mirrenderergl

Reviewer: _I_ wouldn't expand mirserver here. I think that introduces too many concepts.

Reviewer: Sure, they need explaining but we're well past the number of concepts that should be introduced together. (Even Agatha Christie stops at a dozen suspects and she was aiming to confuse!)

First introduce [the high level concepts](https://github.com/MirServer/mir/pull/3053#pullrequestreview-1650985226). Then, break down each one in turn: (e.g. "These are the platforms... this one is for X11...", "This is the mirserver... it has these components...").

That gives the reader some structure to organise their understanding.

├── mirplatform
├── mirwayland
├── graphics-gbm-kms
├── graphics-eglstreamkms
├── graphics-x11
├── graphics-wayland
├── input-evdev
└── renderer-egl-generic
```

## miral
- The "Mir Abstraction Layer" (aka `MirAL` or `miral`) is an API that makes Mir easy for compositor-implementers to work with. It provides the core window management concepts and an interface that is more ABI stable than Mir's internal mirserver API.
- Public-facing
- Directories:
  - `include/miral`
  - `src/miral`
  
## miroil
- A custom *Mir* library for the [Lomiri](https://lomiri.com/) folks
- This is unimportant for 99% of people
- Directories:
  - `include/miroil`
  - `src/miroil`

## mircore
- Fundamental data structures like 2D points, Rectangles, File Descriptor, etc.
- Public-facing
- Directories:
  - `include/core`
  - `src/core`
  -
## mircommon
- Data structures and utility functions like Mir-event building, logging, keymaps, etc.
- Differs from `mircore` in that the concepts map strictly onto *Mir*'s concepts instead of being general concepts like a Rectangle
- Public-facing
- Directories:
  - `include/common`
  - `src/common`
  - `src/include/common`

## mircookie

> - Provides event timestamps for inputs events that are difficult to spoof.

Reviewer: @RAOF I suspect that mircookie isn't actually useful now we're no longer shipping Mir's input events to clients. WDYT?

- Public-facing
- Directories:
  - `include/cookie`
  - `src/cookie`
  - `src/include/cookie`

## mirserver
- Provides the core *Mir* runtime, with the entry-point being the `mir::Server` class.
- The `Server` class has a `DisplayServer` which provides access to the interfaces that that support the server's runtime.
  - The `ServerConfiguration` provides a bunch of methods called `the_XYZ()` to access these various instances, where "XYZ" can be replaced with the interface of interest.
- The classes that the `Server` relies on for its runtime can be found organized in separate folders in this directory.
- Directories:
  - `include/server`
  - `src/server`
  - `src/include/server`
  

> ## mircompositor

Reviewer: I don't like this naming (probably comes from trying to treat it at the same level of abstraction as libraries).

It is "just" a namespace within mirserver.

Author: It could be `mirserver::mircompositor`?

Reviewer: That's worse! How is a reader to associate something in the code (`mir::compositor::Foo`, say) with `mirserver::mircompositor`?

- Included in `mirserver.so`
- The `Compositor` is responsible for collecting the set of renderables and sending them off to the outputs to be displayed.
- Directories:
  - `src/include/server/mir/compositor`
  - `src/server/compositor`

## mirinput
- Included in `mirserver.so`
- Deals with everything input (e.g. keyboard presses, mouse clicks)
- The `InputManager` provides access to the specific input platform that is running (e.g. X, Wayland, or libinput).
- The `InputDispatcher` is responsible for taking events bubbled up from the platform and sending them off to the appropriate listeners
- Directories
  - `src/include/server/mir/input`
  - `src/server/input`
  
## mirfrontend

> - TODO: How do I describe this one?

Reviewer: `src/server/frontend` is vestigial - we should probably merge it with `src/server/frontend_wayland`

- Directories:
  - `src/include/server/mir/frontend`
  - `src/server/frontend`
  
## mirfrontend-wayland
- The *Mir* implementation of the wayland protocol.
- The `WaylandConnector` class is the glue between the compositor and the specific details of the wayland protocol.
- Directories:
  - `src/server/frontend_wayland`

## mirfrontend-xwayland
- The *Mir* implementation of the xwayland protocol.
- The `XWaylandConnector` class is the glue between the compositor and the specific details of the xwayland protocol.
- This protocol talks to the `xwayland` server, which is spawned as a subprocess.
- Directories:
  - `src/server/frontend_xwayland`

## mirshell
- The `Shell` is responsible for managing the surfaces that the compositor wants to show.
- It provides an interface to create, update, and remove these surfaces from the display.
- As an example, the `WaylandConnector` might ask the `Shell` to create a new surface for it.
- Directories:
  - `src/include/server/mir/shell`
  - `src/server/shell`
  
## mirshelldecoration

> - Manages the decoration of *Mir*'s windows

Reviewer: ```suggestion
- Manages *Mir*'s server-side decoration of windows. Currently only used for X11 clients.
```

- Directories:
  - `src/server/shell/decoration`

## mirscene
- The `Scene` provides an interface for the `Compositor` to access the list of renderable items
  - These renderables are derived from the surfaces that were added to the `Shell`.
  - You can think of the `Scene` as you would think of a "scene graph" in a 3D game engine.
- Unlike the `Shell`, the `Scene` knows a lot about the Z-order of the surfaces. For this reason, it is also responsible for things like locking the display or sending surfaces to the back of the Z-order.
- Directories:
  - `src/include/mir/server/scene`
  - `src/server/scene`

## mirgraphics
- A graphics abstraction layer sitting between the compositor and the specific graphics platform that it is using to render.
- Directories:
  - `src/include/mir/server/graphics`
  - `src/server/graphics`
  
## mirreport
- Default logging and reporting facilities
- Directories:
  - `src/server/report`

> ## mirlttng

Reviewer: I like this naming even less. `lttng` is a directory/namespace within `report` - it shouldn't be named at the same level as `mirwayland`. Similarly, `logging`, `null`.

- Handles `logind` and virtual-terminal related tasks for the compositor
- Directories:
  - `src/server/console`

## mirgl
- A short list of helpers for GL things
- Directories:
  - `src/gl`
  
## mirrenderergl
- The only supported `Renderer` type for now
- Future renderers will be found alongside it in the `src/renderers` directory
- Directories:
  - `src/renderers/gl`
  
## mirplatform
- Graphics and input code that is shared between the platforms found in `src/platforms`
- Directories
  - `src/include/platform`
  - `src/platform`

## mirwayland
- A subproject used to generate C++ classes from wayland protocol XML files.
- Directories:
  - `src/wayland`

## graphics-gbm-kms
- GBM/KMS display platform for Mir
- This is the main display platform that one would expect to use
- Directories:
  - `src/platforms/gbm-kms`

## graphics-gbm-kms
- EglStreams/KMS display platform for Mir
- This is an Nvidia-specific platform
- Directories:
  - `src/platforms/eglstreams-kms`

## graphics-x11
- An X11-based display/input platform
- Primary platform used for testing
- Directories:
  - `src/platforms/x11`
  
## graphics-wayland
- A Wayland-based display/input platform
- Directories:

>   - `src/platforms/wayland`

Reviewer: Suddenly we're not prefixing everything with "mir"! :grinning: 

I think it works better to explain what a platform is, and then list examples.

Author: I believe I only did this because the final `.so` names aren't prefaced with mir here :joy: But yes, agreed

### Miroil
`miroil` is a custom library for the [Lomiri](https://lomiri.com/) folks. Most compositor authors will not interact with `miroil`. [The *Mir* engine](#the-mir-engine) also serves as the basis for `miroil` as it does for `miral`.

### Mirwayland
Compositor authors may want to define their own wayland protocol extensions in addition to the ones that the core *Mir* implementation defines. The `mirwayland` library satisfies this requirement. This library may be used in conjunction with either `miral` or `miroil`.

## The Mir engine
The `mirserver` library is the core implementation of *Mir*. It serves as the engine for both `miral` and `miroil`. This library does the heavy-lifting for the compositor and, as a result, it is the largest piece of *Mir*. This section will explain the primary concepts that drive the engine.

### Core Concepts

> At the heart of `mirserver` are two interfaces: the **Shell** and the **Scene**. The `Shell` is responsible for fielding requests from the rest system. The `Shell` then modifies the state of the `Scene` to reflect the requested changes. For example, part of the system would ask the `Shell` to move a window. The `Shell` would then decide what position to move that window to and update the state of the `Scene` to reflect that change.

Reviewer: ```suggestion
At the heart of `mirserver` are two interfaces: the **Shell** and the **Scene**. The `Shell` is responsible for fielding requests from the rest system. The `Shell` then modifies the state of the `Scene` to reflect the requested changes. For example, the `Frontend` would ask the `Shell` to initiate dragging a window. The `Shell` would then decide how to move that window to and update the state of the `Scene` to reflect that change.
```
